### PR TITLE
[Test Generation] Control inlining using method annotations

### LINKED ISSUE #3716: Allow annotating methods/functions to control inlining for test generation
### Summary

Have an annotation or option to mark a given method as "to be covered" if it is reachable in the call graph of the entry point for test generation.

### Background and Motivation

Currently one way to control the coverage requirements of the test generation is to use the `--inline-depth n` argument, which asks for coverage of all the blocks of code within methods at most `n` steps away from the entry point in the call-graph. However there is a strong potential here for path explosion, the usual problem encountered with this kind of symbolic test generation.

Ideally things would be a bit more fine-grained, by marking exactly which methods should be covered, if reachable from the entry point.

### Proposed Feature

One could imagine either a command line feature like `--inline-for-tests SomeModule.SomeMethod`, or an attribute `{:test-inline}` attached to the method/function method declaration itself.

I can imagine 2 semantics for this annotation:

- The method, *and all methods above it in the call graph* are inlined, and the test generation tool attempts to cover each block of the annotated method using the usual approach.
- The method is inlined iff all the methods above it are also marked as inlined (or if it is directly reachable in 1 step from the entry point).

The second semantics seems more "honest" in terms of the semantics, but may lead to confusing behavior if the method is "hidden" by a non-inlined method. Maybe a warning is appropriate in this case.

### Alternatives

I don't currently have a case where this feature allows test generation to succeed (i.e. both terminate and cover the desired blocks) when it otherwise would fail, though I suspect I will soon run into this. I'm not sure what an alternative would be.

### PR DESCRIPTION
**Pull request by Dargones**
Fixes #3716

This is a second attempt at adding more fine grained control over inlining to test generation. In lieu of a command line option (as in #3728), this PR introduces a `:testInline` attribute which can be put on methods and functions. The attribute must be followed by a positive integer, which in the presence of recursion gives the bound on how many times the given method can be unrolled (1 means inlining but no unrolling). In addition to using the attribute, one must specify the `--inline` boolean command line option to activate inlining (because inlining requires changes to the parsing process which must happen before I can look at method attributes). All of this very closely mirrors Boogie's API for inlining, so I think it should be more intuitive than the previous solution.

@codyroux, @hmijail, @aleks-AWS, I would be grateful for any feedback on how well this addresses the needs of various applications since there are various trade-off involved in-between this and #3728! Below I discuss some of the details and give an example.

I added various error messages that the user gets if they do not specify the unrolling limit or if they use the `:testInline` attribute but not the `--inline` command line option. Using `--inline` with `--verbose` will also output the list of methods/functions that were inlined, so the user can double-check that the desired methods are in-fact being inlined.

One of the alternatives would be to make inlining behavior specific to the target tested method, as @keyboardDrummer proposes in the linked issue. My guess is that the use case for this is relatively uncommon as there is usually one specific high-level method that has to be targeted but it is true that the solution proposed in this PR is more restrictive. Is there an application that would require different inlining behavior based on what method is being tested?

As an example of how inlining would behave if this PR gets merged, consider the following Dafny program that recursively computes the remainder of dividing some number by `3` (this test case is added as part of the PR):

```Dafny
module M {
  class Inlining {
    function {:testInline 2} mod3 (n:int):int 
      requires n >= 0
      decreases n
    {
      if n == 0 then 0 else
      if n == 1 then 1 else
      if n == 2 then 2 else
      mod3(n-3)
    }
    method test(n:int) returns (r:int) 
      requires n >= 3
    {
      r := mod3(n);
    }
  }
}
```

To generate tests for this program, one could execute:
```bash
dafny generate-tests Block --inline --target-method:M.Inlining.test FILENAME.dfy --verbose
```

Because the value of `:testInline` is `2`, test generation would produce at least `3` tests, one for each of the base cases in the recursive `mod3` function. With `:testInline` set to `1`, test generation would not be able to see beyond the first call to `mod3`, and so it would not be able to generate tests for the base cases as they are effectively unreachable without a recursive call.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Comment by keyboardDrummer** [comment]
> One of the alternatives would be to make inlining behavior specific to the target tested method, as @keyboardDrummer proposes in the linked issue. My guess is that the use case for this is relatively uncommon as there is usually one specific high-level method that has to be targeted but it is true that the solution proposed in this PR is more restrictive. Is there an application that would require different inlining behavior based on what method is being tested?

That could be a future feature instead of an alternative, `{:testInline M.Inlining 3}` on the root method could override the `{:testInline 2}` on the `M.Inlining` method.

**Comment by keyboardDrummer** [comment]
> (because inlining requires changes to the parsing process which must happen before I can look at method attributes)

I'm not sure which parsing process you're referring to. Looking at where `TestGenOptions.TestInline` is used, that seems to be way after parsing the Dafny program.

File: Source/DafnyCore/TestGenerationOptions.cs [hunk 1]
```diff
@@ -52,11 +53,8 @@ public enum Modes { None, Block, Path };
           }
           return true;
 
-        case "generateTestInlineDepth":
-          var depth = 0;
-          if (ps.GetIntArgument(ref depth)) {
-            TestInlineDepth = (uint)depth;
-          }
+        case "generateTestInline":
```
Comment (keyboardDrummer) [inline Source/DafnyCore/TestGenerationOptions.cs:56]: Consider using `DafnyOptions.RegisterLegacyUi` to replace the `generateTestInline` related code in this file.

**Review by keyboardDrummer — COMMENTED** [review]
Looks great, left some minor comments

**Comment by Dargones** [comment]
Thank you, @keyboardDrummer for all the feedback! I have removed the `--inline` option like you suggest above (meaning that I parse the input program, scan it for the presence of the `:testInline` attribute and then re-parse it again if the attribute is present). The re-parsing is necessary because I need to look at the bodies of all inlined methods/functions, even if they are defined in a different file (so I have to parse with `/verifyAllModules` flag).

I think this should also address your comments about how to register the command line option properly (since I removed the option altogether), but I am wondering if I should make a separate PR to make the other options use the `DafnyOptions.RegisterLegacyUi` functionality. Would you recommend creating a `TestGenerationOptionsBag` file for this purpose?

File: Source/DafnyTestGeneration/GenerateTestsCommand.cs [hunk 2]
```diff
@@ -84,8 +84,11 @@ private enum Mode {
     DafnyOptions.RegisterLegacyBinding(SequenceLengthLimit, (options, value) => {
       options.TestGenOptions.SeqLengthLimit = value;
     });
-    DafnyOptions.RegisterLegacyBinding(TestInlineDepth, (options, value) => {
-      options.TestGenOptions.TestInlineDepth = value;
+    DafnyOptions.RegisterLegacyBinding(TestInline, (options, value) => {
```
[the remark this answers is not in the document]
Comment (Dargones) [reply Source/DafnyTestGeneration/GenerateTestsCommand.cs]: I have decided to remove this option completely and just scan the program for the `:testInline` attribute and reparse if necessary (see my comment below), but thank you for clarifying how to register the options properly, this is very helpful!

**Comment by Dargones** [comment]
Have made a commit that ensures the transformations I apply to support inlining only modify procedures that are going to be inlined - this should significantly improve test generation performance.

**Review by keyboardDrummer — APPROVED** [review]

**Comment by atomb** [comment]
It looks as though some code changed on `master` removed or moved a field this PR is using.

**Comment by Dargones** [comment]
Thank you, @atomb! Should work now

**Review by atomb — APPROVED** [review]

**Pull request merged**